clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1236,7 +1243,17 @@ public void assign(Collection<TopicPartition> partitions) {
                 }
 
                 final FetchedRecords<K, V> records = pollForFetches(timer);
-                if (!records.isEmpty()) {
+
+                // We only need to save off the metadata if we are NOT supposed to return early on metadata-only responses
+                if (!longPollExitEarlyOnMetadata) {
```

Comment:
How about updating `nullableSeenMetadata` only if we don't return records?
```java
if (longPollShouldReturn(records)) { 
 ...
} else {
  if (nullableSeenMetadata == null) {
    nullableSeenMetadata = new HashMap<>(records.metadata());
  } else {
    nullableSeenMetadata.putAll(records.metadata());
  }
}
```

The benefit from above code is that we don't need to handle duplicate metadata which exists on both `FetchedRecords` and `nullableSeenMetadata` when it succeed to get records and metadata in first loop.

Reply:
sounds good!